### PR TITLE
Line search

### PR DESCRIPTION
**Pull request by RikVoorhaar**
I made a Riemannian line search optimizer with strong Wolfe conditions. 

It's not yet perfect. I think it makes some redundant calls to `closure` during stepping, and when it's close to a local minimum it suffers from numerical errors and can sometimes take strange steps

### CODE REVIEW

File: geoopt/optim/rwolfe.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+            fallback_stepsize=fallback_stepsize,
+            amax=amax,
+            amin=amin,
+            stabilize=stabilize,
+        )
+        super(RiemannianLineSearch, self).__init__(
+            params, defaults, stabilize=stabilize
+        )
+        self._params = self.param_groups[0]["params"]
+        self.c1 = self.param_groups[0]["c1"]
```
Comment (ferrine) [inline geoopt/optim/rwolfe.py:77]: why is it global? Is it possible to have different constants between param groups?

File: geoopt/optim/rwolfe.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+        self.phi_(step_size)
+
+        derphi = 0
+        for point in self._params:
+            state = self.state[point]
+            if "grad" not in state:
+                continue
+
+            # For some reason using the metric for inner product gives wrong results,
+            # therefore us euclidean product instead.
```
Comment (ferrine) [inline geoopt/optim/rwolfe.py:161]: what was wrong there?

File: geoopt/optim/rwolfe.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+                continue
+            if isinstance(point, (ManifoldParameter, ManifoldTensor)):
+                manifold = point.manifold
+            else:  # Use euclidean manifold
+                manifold = self._default_manifold
+
+            state = self.state[point]
+            # project gradient onto tangent space
+            grad = manifold.egrad2rgrad(point, grad)
+            grad_norms.append(grad.norm().item())
```
Comment (ferrine) [inline geoopt/optim/rwolfe.py:189]: do you really need Euclidean norm here and not Riemannian one?

File: geoopt/optim/rwolfe.py (hunk 1 above)
Comment (RikVoorhaar) [reply geoopt/optim/rwolfe.py:77]: `stabilize` should probably be per parameter group, but I don't see a way to implement different line search constants for different parameter groups.

File: geoopt/optim/rwolfe.py (hunk 2 above)
Comment (RikVoorhaar) [reply geoopt/optim/rwolfe.py:161]: I now realized what went wrong probably. If using `manifold.inner(point, -state["new_grad"], state["grad_retr"])` at this point in the code, then the point is wrong. We need to compute this quantity before we roll back point at the end of the function `_phi`, and doing this seems to work well. I will modify the code.

File: geoopt/optim/rwolfe.py (hunk 3 above)
Comment (RikVoorhaar) [reply geoopt/optim/rwolfe.py:189]: No, Riemannian is better, thank you for the observation. I will update this.

File: geoopt/optim/rwolfe.py (hunk 1 above)
Comment (ferrine) [reply geoopt/optim/rwolfe.py:77]: > `stabilize` should probably be per parameter group

Fair point, I missed that

Comment (ferrine) [reply geoopt/optim/rwolfe.py:77]: > I don't see a way to implement different line search

That means to avoid misunderstanding between user and developer we need some checks for that or do not set is per parameter group in the first place

File: geoopt/optim/rwolfe.py (hunk 3 above)
Comment (ferrine) [reply geoopt/optim/rwolfe.py:189]: btw, all norm computations are broadcastable and can factorize across product manifold. if you have a product manifold, you may better have more than one grad norm. 

```
# pseudocode
point = Sphere(10, 3)  # 10 3d spheres
inner = point.inner(v1, v2)
inner.shape
(10,)
```

Does having a factorized inner product makes sense in this optimizer? In Adam this affected adaptive term in a good way

Comment (RikVoorhaar) [reply geoopt/optim/rwolfe.py:189]: I just need the squared sum of gradient norms to perform the line search, so this isn't necessary. But this is good to keep in mind if implementing geometric conjugate gradient (which seems useful).

File: geoopt/optim/rwolfe.py [hunk 4]
```diff
@@ -186,12 +190,11 @@ def init_loss(self):
             state = self.state[point]
             # project gradient onto tangent space
             grad = manifold.egrad2rgrad(point, grad)
-            grad_norms.append(grad.norm().item())
+            grad_norm = manifold.norm(point, grad)
```
Comment (ferrine) [inline geoopt/optim/rwolfe.py:193]: Be careful, norm and inner reduce only manifold dimensions, not batch ones. Was it what you meant?

File: geoopt/optim/rwolfe.py [hunk 5]
```diff
@@ -130,6 +138,8 @@ def phi_(self, step_size):
 
             state["new_grad"] = grad
 
+            state["der_phi"] = manifold.inner(point, -grad, state["grad_retr"])
```
Comment (ferrine) [inline geoopt/optim/rwolfe.py:141]: Same as comment below about reduction dimensions, you better use sum after calling norm

File: geoopt/optim/rwolfe.py (hunk 4 above)
Comment (RikVoorhaar) [reply geoopt/optim/rwolfe.py:193]: Good point. I did not design this optimizer with batches in mind (e.g. line search doesn't really make sense in the stochastic gradient descent setting). I will alter the optimizer so that it works correctly with batches of tensors.

**Comment by ferrine** [comment]
I meant batches of tensors that are product manifolds

**Comment by RikVoorhaar** [comment]
> I meant batches of tensors that are product manifolds

In that case I don't think I understand what you mean. Without modification it seems to work just fine with tensors belonging to a `ProductManifold`. Either way, I modified the code to use `sum` in the cases you mentioned, and it now seems to work just fine with batches of tensors as well.

**Comment by RikVoorhaar** [comment]
I figured that adding line search based on Armijo backtracking is basically free, so I added this functionality. 
Might also add weak wolfe conditions. Adding support for weak Wolfe conditions, and Hager-Zhang line search would also be nice, but requires more effort. Same goes for implementing conjugate gradient methods.

**Comment by ferrine** [comment]
You mentioned 
> when it's close to a local minimum it suffers from numerical errors and can sometimes take strange steps

Is this problem investigated?

File: geoopt/optim/rwolfe.py [hunk 6]
```diff
@@ -0,0 +1,280 @@
...
+                continue
+            if isinstance(point, (ManifoldParameter, ManifoldTensor)):
+                manifold = point.manifold
+            else:  # Use euclidean manifold
+                manifold = self._default_manifold
+
+            state = self.state[point]
+            # project gradient onto tangent space
+            grad = manifold.egrad2rgrad(point, grad)
+            grad_norm = torch.sum(manifold.norm(point, grad)).item()
```
Comment (ferrine) [inline geoopt/optim/rwolfe.py:206]: Norm here is computed incorrect, I guess you meant
```suggestion
            grad_norm = torch.sum(manifold.inner(point, grad)).pow(.5).item()
```

File: geoopt/optim/rwolfe.py [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+                manifold = self._default_manifold
+
+            state = self.state[point]
+
+            # project gradient onto tangent space
+            grad = manifold.egrad2rgrad(point, grad)
+
+            state["new_grad"] = grad
+
+            state["der_phi"] = torch.sum(
```
Comment (ferrine) [inline geoopt/optim/rwolfe.py:152]: btw, what happens in case you have more than one parameter? In here you compute norms, etc, per parameter. But from the conversation, I felt like you need a global norm, global inner product

File: geoopt/optim/rwolfe.py (hunk 6 above)
Comment (RikVoorhaar) [reply geoopt/optim/rwolfe.py:206]: > Norm here is computed incorrect, I guess you meant

what is the difference between the two? Shouldn't manifold.norm do exactly this?

File: geoopt/optim/rwolfe.py (hunk 7 above)
Comment (RikVoorhaar) [reply geoopt/optim/rwolfe.py:152]: The `state["new_grad"]` isn't used, but should be. If the step we take is the one suggested by the line search, then `state["new_grad"]` already contains the gradient of the loss function with respect to this parameter, and we can omit a call to closure(). I had this idea in mind, but forgot to implement it. This does assume that the user does not alter the parameters in any way between consecutive calls of `.step()`, and I'm not whether such an assumption is necessarily good. 

`state["der_phi"]` is used in the function `derphi_`, where we just sum over all of the paremeters' `state["der_phi"]`. This sum is the global inner product we need. So, right, in principle we do not need to store it like this and we can just compute and store the sum from the get-go, but I figured storing this might not be bad since it tells us the contribution of a particular parameter to the gradient of the line search objective, which could be a useful metric for the user. (though this should be documented).

**Comment by RikVoorhaar** [comment]
> You mentioned
> 
> > when it's close to a local minimum it suffers from numerical errors and can sometimes take strange steps
> 
> Is this problem investigated?

Yes, if this line search fails to find a new step it falls back to a step of a fixed size, which tends to be very large step. 
This should only happen in case the point is near a local minimum (or saddle point). Typically this might be the point where we halt the optimization process (or before this point already). Perhaps a good option would be to default to not doing a step at all, instead of using a large fallback step size, and simply warn the user that the convergence has been reached and additional steps are prone to numerical error.

File: geoopt/optim/rwolfe.py (hunk 7 above)
Comment (ferrine) [reply geoopt/optim/rwolfe.py:152]: > This does assume that the user does not alter the parameters in any way between consecutive calls 

I believe this covers 99.9% usecases. This is also the assumption in all geoopt optimizers that have momentum support.

**Comment by ferrine** [comment]
I

> > You mentioned
> > > when it's close to a local minimum it suffers from numerical errors and can sometimes take strange steps
> > 
> > 
> > Is this problem investigated?
> 
> Yes, if this line search fails to find a new step it falls back to a step of a fixed size, which tends to be very large step.
> This should only happen in case the point is near a local minimum (or saddle point). Typically this might be the point where we halt the optimization process (or before this point already). Perhaps a good option would be to default to not doing a step at all, instead of using a large fallback step size, and simply warn the user that the convergence has been reached and additional steps are prone to numerical error.

Hmm, makes sense. Do you know any practical way to detect convergence of line search to implement here? Halting by default sounds good

**Comment by RikVoorhaar** [comment]
> Hmm, makes sense. Do you know any practical way to detect convergence of line search to implement here? Halting by default sounds good
Convergence happens if the gradient of the line search objective is small (since this is nothing but the norm-squared of the function gradient). Unfortunately "small" is not a well-defined concept, so checking whether or not the optimization has converged is up to the user. From my experiments, the line search fails to find a new point mainly when it is close to a local minimum. 
I left it now that by default the optimizer does not take a step if the it can't find a good stepsize, but the user can call the `step` method with `force_step=True` to override this behavior.

**Comment by ferrine** [comment]
There was a strange failure in travis, restarting build

**Comment by ferrine** [comment]
oh it's doctest failing

**Comment by RikVoorhaar** [comment]
> oh it's doctest failing

yes, I noticed this, but the error seemed too trivial to fix with a new commit.

**Comment by RikVoorhaar** [comment]
travis errors are in `test_gyrovector_math.py` and `test_lorentz_math.py`, was getting the same thing locally.

**Comment by ferrine** [comment]
Looks like pytest update

**Comment by RikVoorhaar** [comment]
I'm now thinking of partially rewriting this class for more general behavior. 

Right now search direction of line search is fixed to be gradient direction. Instead we can use conjugate gradient search direction:

CodeCogsEqn

[(Zhu-Sato suggests to use inverse retraction here instead of vector transport)](http://www.optimization-online.org/DB_FILE/2020/05/7798.pdf)

Steepest descent then boils down to setting beta = 0. Most popular choices of beta_{k+1} involve the same parameters:

CodeCogsEqn (1)

These parameters are all typically computed during the line search anyway, so it's free to use this. The idea is then to make all functions determining beta_{k+1} use the same API, so that it's easy for the user to write new functions. 

There are also many different ways of doing the line search itself, and there are many stopping criteria for the linesearch possible. Even in some particular cases exact line search is feasible. Right now it's not simple for the user to change the linesearch method to something new, so the class should be rewritten to make linesearch methods use the same API.

All in all, I think these modifications will make the optimizer more useful and general.

**Comment by ferrine** [comment]
I'll fix pytest soon in master

**Comment by ferrine** [comment]
You can rebase to fix pytest errors

**Comment by RikVoorhaar** [comment]
One of tests is failing only on MacOS. This test is a bit prone to numerical errors, so I need to implement a better test. I have a list of test problems, most of which are less prone to numerical errors.

File: geoopt/optim/rwolfe.py [hunk 8]
```diff
@@ -0,0 +1,689 @@
...
+
+            # Compute retract if suggested step size is not the last one we tried (rare)
+            if redo_retract:
+                new_point = manifold.retr(point, step_size * search_direction)
+            else:
+                new_point = state["new_point"]
+
+            # Use retract to perform the step, and transport the search direction
+            if self.transport_search_direction:
+                search_transported = manifold.transp_follow_retr(
```
Comment (RikVoorhaar) [inline geoopt/optim/rwolfe.py:502]: Using `transp_follow_retr` here seems less efficient in general than using `transp`.  However, for CanonicalStiefel for example, `transp` is not defined but `transp_follow_retr` is.

**Comment by RikVoorhaar** [comment]
I added conjugate gradients and made it easier for the user to supply their own linesearch functions. I also fixed a number of small bugs, and found a couple points where I could make it a bit more efficient.

**Comment by RikVoorhaar** [comment]
Fletcher Reeves scale parameter for conjugate gradient method seems to not work well with the orthogonal procrustes problem I use for testing. I need to investigate more carefully if there is no error in the implementation.

**Comment by RikVoorhaar** [comment]
It seems the problem is solved by taking square norm as loss in the test, instead of just norm. In fact before the gradient norms didn't even go to zero.

**Comment by ferrine** [comment]
Hi! Is it time for review or you have any ideas to improve docs?

**Comment by RikVoorhaar** [comment]
I made some minor changes, most importantly I renamed rwolfe.py to rlinesearch.py, since this name is more appropriate (in fact it seems that in most practical cases armijo outperforms wolfe line search in the Riemannian setting). 

It should be ready for review. I don't know how to improve the docs any further.

**Comment by ferrine** [comment]
I'll have look this week! Thanks!

**Review by ferrine — COMMENTED** [review]
Looks good, you just need to merge master in this branch. And we are ready

**Comment by RikVoorhaar** [comment]
done

**Comment by ferrine** [comment]
Could you please fix the rest ?  There seem to be some unused imports in tests

**Comment by RikVoorhaar** [comment]
should be good now

**Comment by ferrine** [comment]
Great, thanks, that is solid
